bastion/src/path.rs

```diff
@@ -0,0 +1,459 @@
+    pub(crate) fn root() -> BastionPath {
+        BastionPath {
+            parent_chain: vec![],
+            this: None,
+        }
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = &BastionId> {
+        let parent_iter = self.parent_chain.iter();
+        parent_iter.chain(self.this.iter().map(|e| e.id()))
+    }
+
+    pub(crate) fn id(&self) -> &BastionId {
+        self.this.as_ref().map(|e| e.id()).unwrap_or(&NIL_ID)
+    }
+
+    pub fn elem(&self) -> &Option<BastionPathElement> {
+        &self.this
+    }
+}
+
+impl fmt::Display for BastionPath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "/{}",
+            self.iter()
+                .map(|id| format!("{}", id))
+                .collect::<Vec<String>>()
+                .join("/")
+        )
+    }
+}
+
+impl fmt::Debug for BastionPath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.this {
+            Some(this @ BastionPathElement::Supervisor(_)) => write!(
+                f,
+                "/{}",
+                self.parent_chain
+                    .iter()
+                    .map(|id| BastionPathElement::Supervisor(id.clone()))
+                    .chain(vec![this.clone()])
+                    .map(|el| format!("{:?}", el))
+                    .collect::<Vec<String>>()
+                    .join("/")
+            ),
+            // TODO: combine with the pattern above when or-patterns become stable
+            Some(this @ BastionPathElement::Children(_)) => write!(
+                f,
+                "/{}",
+                self.parent_chain
+                    .iter()
+                    .map(|id| BastionPathElement::Supervisor(id.clone()))
+                    .chain(vec![this.clone()])
+                    .map(|el| format!("{:?}", el))
+                    .collect::<Vec<String>>()
+                    .join("/")
+            ),
+            Some(this @ BastionPathElement::Child(_)) => {
+                let parent_len = self.parent_chain.len();
+
+                write!(
+                    f,
+                    "/{}",
+                    self.parent_chain
+                        .iter()
+                        .enumerate()
+                        .map(|(i, id)| {
+                            if i == parent_len - 1 {
+                                BastionPathElement::Children(id.clone())
+                            } else {
+                                BastionPathElement::Supervisor(id.clone())
+                            }
+                        })
+                        .chain(vec![this.clone()])
+                        .map(|el| format!("{:?}", el))
+                        .collect::<Vec<String>>()
+                        .join("/")
+                )
+            }
+            None => write!(f, "/"),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum BastionPathElement {
```

Comment:
Same applies for here.

Reply:
I've written examples for methods which are most likely will be called by a public api user.
For structs I've described what they are representing but without examples because at this point there is no public api for constructing this structs.

Reply:
Generally speaking, we want everything that is public to either be documented and have an example or be `#[doc(hidden)]`